### PR TITLE
keepalive: apply minimum ping time of 10s to client and 1s to server

### PR DESCRIPTION
**Pull request by dfawley**
Also fix tests that relied upon setting lower values.

This is in preparation for the fix to issue #2638

### CODE REVIEW

File: keepalive/keepalive.go [hunk 1]
```diff
@@ -62,6 +63,7 @@ type ServerParameters struct {
 	MaxConnectionAgeGrace time.Duration // The current default value is infinity.
 	// After a duration of this time if the server doesn't see any activity it
 	// pings the client to see if the transport is still alive.
+	// If set below 1s, a minimum value of 1s will be used instead.
```
[the remark this answers is not in the document]
Comment (dfawley) [reply keepalive/keepalive.go]: That's how the spec is.  The server has no minimum, but I'm thinking of using 1s granularity on the ping timer, so 1s would be the practical minimum.

File: clientconn_test.go [hunk 2]
```diff
@@ -939,25 +939,60 @@ func (s) TestClientUpdatesParamsAfterGoAway(t *testing.T) {
...
+		conn, err := lis.Accept()
+		if err != nil {
+			t.Errorf("error accepting connection: %v", err)
+			return
+		}
+		defer conn.Close()
+		f := http2.NewFramer(conn, conn)
+		go func() {
+			for {
+				fr, err := f.ReadFrame()
```
[the remark this answers is not in the document]
Comment (dfawley) [reply clientconn_test.go]: Fixed.  (I was logging this earlier while debugging.)

File: test/channelz_test.go [hunk 3]
```diff
@@ -847,8 +848,8 @@ func doIdleCallToInvokeKeepAlive(tc testpb.TestServiceClient, t *testing.T) {
 	if err != nil {
 		t.Fatalf("TestService/FullDuplexCall(_) = _, %v, want <nil>", err)
 	}
-	// 2500ms allow for 2 keepalives (1000ms per round trip)
-	time.Sleep(2500 * time.Millisecond)
+	// Allow for at least 2 keepalives (1s per ping interval)
+	time.Sleep(4000 * time.Millisecond)
```
[the remark this answers is not in the document]
Comment (dfawley) [reply test/channelz_test.go]: Done

File: server.go [hunk 4]
```diff
@@ -182,6 +182,11 @@ func InitialConnWindowSize(s int32) ServerOption {
 
 // KeepaliveParams returns a ServerOption that sets keepalive and max-age parameters for the server.
 func KeepaliveParams(kp keepalive.ServerParameters) ServerOption {
+	if kp.Time > 0 && kp.Time < time.Second {
+		grpclog.Warningf("Adjusting keepalive ping interval to minimum period of 1s")
```
[the remark this answers is not in the document]
Comment (dfawley) [reply server.go]: Done

File: test/channelz_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (dfawley) [reply test/channelz_test.go]: Done

**Comment by JelteF** [comment]
Could you explain why it's necessary to disallow any lower ping than 10s/1s? We've been using a keep alive interval of 100ms successfully for a long time now. We use it detect connection failures quickly. To give you some indication of our setup, it's within one datacenter with a limited number of clients. This change is blocking for us to upgrade to grpc 1.19.0.

**Comment by dfawley** [comment]
@JelteF The 10s client-side limit is a fix to bring us in line with the gRPC spec for keepalives here:

https://github.com/grpc/proposal/blob/master/A8-client-side-keepalive.md#extending-for-basic-health-checking

Server-side, there is no minimum defined in the spec.  However, we intend to implement second-level precision as part of a fix for #2638, which would mean sub-second times could not be honored.

**Comment by JelteF** [comment]
Thanks for the quick response. To clear up one thing, we use 100ms as the client keep alive. This is to very quickly detect a broken server process so we can failover to another server.

It makes sense to follow the spec better, but in this case it would be really nice if the minimum keepalive could be configured by the library user as well. Especially since lower keep alives were possible before. And also because the server already has the minimum time enforcement policy.

**Comment by dfawley** [comment]
@JelteF We could potentially allow a lower limit to be configured, given that we had unintentional support for this in the past.  If this is a valid use case, however, we should think about how we would support similar scenarios in Java and C, which currently also restrict clients to 10s keepalive ping intervals.  We have another client-side health checking mechanism ([gRFC](https://github.com/grpc/proposal/blob/master/A17-client-side-health-checking.md)), but that will only quickly detect a problem where the server is still alive but reports itself as unhealthy, not in the case where the server becomes unavailable.

cc @ejona86 who designed client-side keepalives.  He had concerns about this being used to detect network outages (where 100ms is probably too low to be reliable), but it sounds like you are mostly concerned about the server itself becoming wedged.

**Comment by ejona86** [comment]
Our servers have a minimum time enforcement policy, but not all do. The worry is that people really like setting "really low" values without understanding the network/load implications. For example, if the server is overloaded, you don't want the client to disconnect and reconnect. It is a dangerous tool to wield at 100ms, as it can amplify problems.

**Pull request merged**